sql/core/src/test/scala/org/apache/spark/sql/SparkSessionBuilderSuite.scala

```diff
@@ -40,6 +45,37 @@ class SparkSessionBuilderSuite extends SparkFunSuite with BeforeAndAfterEach {
     SparkSession.clearDefaultSession()
   }
 
+  test("SPARK-34087: Fix memory leak of ExecutionListenerBus") {
+    val spark = SparkSession.builder()
+      .master("local")
+      .getOrCreate()
+
+    def listenersNum(): Int = {
+      spark.sparkContext
+        .listenerBus
+        .listeners
+        .asScala
+        .count(_.isInstanceOf[ExecutionListenerBus])
+    }
+
+    (1 to 10).foreach { _ =>
+      spark.cloneSession()
+      SparkSession.clearActiveSession()
+    }
+
+    var num = listenersNum()
+    // Before GC, the number of ExecutionListenerBus is (1 + 10)
+    assert(num === 11)
+    eventually(timeout(10.seconds), interval(1.seconds)) {
+      System.gc()
+      num = listenersNum()
+      // After GC, the number of ExecutionListenerBus should be less than 11
+      // Since GC can't 100% guarantee all out-of-referenced objects be cleaned at one time,
+      // here, we check at least one listener is cleaned up to prove the mechanism works.
+      assert(num < 11)
```

Comment:
shall we simply check `assert(num < 10)`? We created 10 sessions and if `num < 10`, it proves the listeners from the session are GCed.

Reply:
We actually create 11 sessions: original one and 10 cloned.